libraries/botbuilder-core/src/turnContext.ts

```diff
@@ -804,26 +808,17 @@ export class TurnContext {
 
     /**
      * @private
+     * Executes `handlers` as a chain, returning a promise that resolves to the final result.
      */
-    private emit<T>(
-        handlers: ((context: TurnContext, arg: T, next: () => Promise<any>) => Promise<any>)[],
-        arg: T,
-        next: () => Promise<any>
-    ): Promise<any> {
-        const list: ((context: TurnContext, arg: T, next: () => Promise<any>) => Promise<any>)[] = handlers.slice();
-        const context: TurnContext = this;
-        function emitNext(i: number): Promise<void> {
-            try {
-                if (i < list.length) {
-                    return Promise.resolve(list[i](context, arg, () => emitNext(i + 1)));
-                }
-
-                return Promise.resolve(next());
-            } catch (err) {
-                return Promise.reject(err);
-            }
-        }
+    private emit<A, T>(
```

Comment:
This is the bulk of the PR: we now have a much more type-safe way of running emit handlers as a chain.